rasa/utils/tensorflow/tf_models.py

```diff
@@ -106,96 +117,169 @@ def fit(
         if not disable:
             logger.info("Finished training.")
 
-    def _get_tf_train_functions(
+    def train_on_batch(
+        self, batch_in: Union[Tuple[np.ndarray], Tuple[tf.Tensor]]
+    ) -> None:
+        """Train on batch"""
+
+        with tf.GradientTape() as tape:
+            total_loss = self._total_batch_loss(batch_in)
+
+        gradients = tape.gradient(total_loss, self.trainable_variables)
+        self._optimizer.apply_gradients(zip(gradients, self.trainable_variables))
+
+    def build_for_predict(
+        self, predict_data: RasaModelData, eager: bool = False
+    ) -> None:
+        def predict_dataset_function(  # to reuse the same helper method
+            _batch_size: Union[tf.Tensor, int]
+        ) -> tf.data.Dataset:
+            return predict_data.as_tf_dataset(_batch_size, "sequence", shuffle=False)
+
+        _, self._predict_function = self._get_tf_functions(
+            predict_dataset_function, self.batch_predict, eager, "prediction"
+        )
+
+    def predict(self, predict_data: RasaModelData) -> Dict[Text, tf.Tensor]:
+        if self._predict_function is None:
+            logger.debug("There is no tensorflow prediction graph")
```

Comment:
The debug method might be misleading. Maybe something like `No tensorflow prediction graph present. Building it now.`

Reply:
Ah, not needed. Will be logged in `_get_tf_functions`. But can you please add a `.` at the end of the log statements. Also for the others further down.